### PR TITLE
WIP futurize stage 2

### PR DESCRIPTION
**Pull request by ryanking**
I ran `futurize -2 -w` and fixed the few tests that failed.

#resolves 1785

### CODE REVIEW

**Comment by ryanking** [comment]
Any insights on the failure? Tests passed locally for me.

**Comment by ryanking** [comment]
Nevermind I was running the tests wrong. I found a few problems I'm working through.

**Comment by ryanking** [comment]
Can anyone give me insight into the failures? My last update fixes all test failures when ran locally.

**Comment by ryanking** [comment]
Has switching to something like travis (travis-ci.org) been discussed?
Sorry if it has and this is an old topic.

-ryan

**Comment by joelarmstrong** [comment]
Yeah, it's definitely in the plan. All of us feel the pain with Jenkins (and our institutional knowledge about debugging/configuring it is gone at this point).

It's not totally straightforward to switch to Travis instantly, though, because some things in the tests still depend on our Jenkins environment. Also, our tests take a long time and we will have to group them sensibly to avoid Travis killing our build after 50 mins. The plan is to switch chunks of tests to Travis piecemeal over the next months.

**Comment by diekhans** [comment]
Why was Jenkins chosen in the first place?

**Comment by mr-c** [comment]
```
/mnt/ephemeral/workspace/toil-pull-request-integration-tests/docs/batchSystem.rst:32: WARNING: autodoc: failed to import class u'AbstractBatchSystem' from module u'toil.batchSystems.abstractBatchSystem'; the following exception was raised:
Traceback (most recent call last):
  File "/mnt/ephemeral/workspace/toil-pull-request-integration-tests/venv/local/lib/python2.7/site-packages/sphinx/ext/autodoc.py", line 551, in import_object
    __import__(self.modname)
  File "/mnt/ephemeral/workspace/toil-pull-request-integration-tests/src/toil/batchSystems/__init__.py", line 16, in <module>
    from past.builtins import cmp
ImportError: No module named past.builtins
```
```
/mnt/ephemeral/workspace/toil-pull-request-integration-tests/docs/jobStore.rst:10: WARNING: autodoc: failed to import class u'AbstractJobStore' from module u'toil.jobStores.abstractJobStore'; the following exception was raised:
Traceback (most recent call last):
  File "/mnt/ephemeral/workspace/toil-pull-request-integration-tests/venv/local/lib/python2.7/site-packages/sphinx/ext/autodoc.py", line 551, in import_object
    __import__(self.modname)
  File "/mnt/ephemeral/workspace/toil-pull-request-integration-tests/src/toil/jobStores/abstractJobStore.py", line 16, in <module>
    from future import standard_library
ImportError: No module named future
```
[...]
```
______________________________ ERROR collecting  _______________________________
venv/local/lib/python2.7/site-packages/py/_path/common.py:372: in visit
    for x in Visitor(fil, rec, ignore, bf, sort).gen(self):
venv/local/lib/python2.7/site-packages/py/_path/common.py:421: in gen
    for p in self.gen(subdir):
venv/local/lib/python2.7/site-packages/py/_path/common.py:421: in gen
    for p in self.gen(subdir):
venv/local/lib/python2.7/site-packages/py/_path/common.py:411: in gen
    if p.check(dir=1) and (rec is None or rec(p))])
venv/local/lib/python2.7/site-packages/_pytest/main.py:643: in _recurse
    ihook = self.gethookproxy(path)
venv/local/lib/python2.7/site-packages/_pytest/main.py:546: in gethookproxy
    my_conftestmodules = pm._getconftestmodules(fspath)
venv/local/lib/python2.7/site-packages/_pytest/config.py:296: in _getconftestmodules
    mod = self._importconftest(conftestpath)
venv/local/lib/python2.7/site-packages/_pytest/config.py:321: in _importconftest
    raise ConftestImportFailure(conftestpath, sys.exc_info())
E   ConftestImportFailure: (local('/mnt/ephemeral/workspace/toil-pull-request-integration-tests/src/toil/batchSystems/mesos/conftest.py'), (<type 'exceptions.ImportError'>, ImportError('No module named past.builtins',), <traceback object at 0x2b67ad5a2cb0>))
```

File: src/toil/jobStores/azureJobStore.py [hunk 1]
```diff
@@ -525,7 +530,7 @@ def _sanitizeTableName(self, tableName):
         This will never cause a collision if uuids are used, but
         otherwise may not be safe.
         """
-        return 'a' + filter(lambda x: x.isalnum(), tableName)
+        return 'a' + [x for x in tableName if x.isalnum()]
```
Comment (mr-c) [inline src/toil/jobStores/azureJobStore.py]:
```
 NonCachingFileStoreTestWithAzureJobStore.testBatchSystemCleanupCanHandleWorkerDeaths 
Traceback (most recent call last):
  File "/mnt/ephemeral/workspace/toil-pull-requests/src/toil/test/src/fileStoreTest.py", line 417, in testBatchSystemCleanupCanHandleWorkerDeaths
    Job.Runner.startToil(A, self.options)
  File "/mnt/ephemeral/workspace/toil-pull-requests/src/toil/job.py", line 750, in startToil
    with Toil(options) as toil:
  File "/mnt/ephemeral/workspace/toil-pull-requests/src/toil/common.py", line 648, in __enter__
    jobStore = self.getJobStore(config.jobStore)
  File "/mnt/ephemeral/workspace/toil-pull-requests/src/toil/common.py", line 781, in getJobStore
    return AzureJobStore(rest)
  File "/mnt/ephemeral/workspace/toil-pull-requests/src/toil/jobStores/azureJobStore.py", line 139, in __init__
    self.namePrefix = self._sanitizeTableName(namePrefix)
  File "/mnt/ephemeral/workspace/toil-pull-requests/src/toil/jobStores/azureJobStore.py", line 533, in _sanitizeTableName
    return 'a' + [x for x in tableName if x.isalnum()]
TypeError: cannot concatenate 'str' and 'list' objects
```

**Comment by ryanking** [comment]
thanks @mr-c I will take a look at these today

**Comment by jessebrennan** [comment]
@ryanking Jenkins has now been opened up so you should be able to view the failures directly.

**Comment by ryanking** [comment]
Thanks @jessebrennan, i can see the errors now. I can't get the integration tests to run locally, so I'm going to try pushing changes here to ge jenkins to run them.

**Comment by jessebrennan** [comment]
Sounds good. You *should* be able to run integration tests locally
(assuming you have an AWS account that you wouldn't mind accruing the
cost). What are the issues you are having?

Also note, Jenkins was having some trouble spinning up slaves and there is
a bit of a backlog of builds to run. It may take longer than normal.

**Comment by ejacox** [comment]
There are AWS problems. The jenkins-it tests are failing. Just look at the other tests for now.

**Comment by ryanking** [comment]
Its possible I was running into the AWS problems.

The first challenge with getting them running is that the neccessary
environment variables are not documented. The docs at
https://toil.readthedocs.io/en/3.9.1/contributing/contributing.html#running-tests
leave out a few things.

**Comment by jessebrennan** [comment]
@ryanking merging in master should resolve the failing integration tests. We pushed a fix for that out quickly yesterday.

The other failing unit tests largely stem from a unicode conversion error in the azureJobStore [here](https://github.com/BD2KGenomics/toil/blob/master/src/toil/jobStores/azureJobStore.py#L473).

**Comment by ryanking** [comment]
@jessebrennan thanks

**Comment by benedictpaten** [comment]
jenkins, test this please

**Comment by ejacox** [comment]
Jenkins, test this please

**Comment by ryanking** [comment]
Sorry for the slow cycles on this - I'm trying to chip away when I have time.

**Comment by ejacox** [comment]
No problem @ryanking . Let us know how we can help.

**Comment by jessebrennan** [comment]
@ryanking I can take on debugging these azure problems. (I was going to start now, but I need the sys admin to help me figure out what my azure account name is...)

**Comment by ryanking** [comment]
Thanks @jessebrennan. It appears that most of the problems so far are handling strings the various string types in the encryption code. It seems that that code is only tested in the integration tests.

**Comment by ejacox** [comment]
@ryanking, was there anything you wanted to add to this PR?

**Comment by ryanking** [comment]
@ejacox nope. 

Wow, thanks for finishing this, I thought I'd have to keep chipping away at this. :)

**Comment by ejacox** [comment]
In the Azure and AWS jobStore code, I convert each filedID argument with bytes() for all calls to the bucket libraries (e.g. through boto). I also tried converting all fileID generation, which are UUIDs, instead. This worked, but it is possible that other fileIDs might come from elsewhere. Therefore, it is probably safer to keep converting the fileIDs with bytes() at the entry points to the underlying libraries.

File: src/toil/jobStores/abstractJobStore.py [hunk 2]
```diff
@@ -32,9 +37,10 @@
 from toil.job import JobException
 from bd2k.util import memoize
 from bd2k.util.objects import abstractclassmethod
+from future.utils import with_metaclass
 
 try:
-    import cPickle
+    import pickle
 except ImportError:
```
[the remark this answers is not in the document]
Comment (joelarmstrong) [reply src/toil/jobStores/abstractJobStore.py]: Nice catch. But this made me notice something. We should really be prioritizing cPickle over pickle in python 2. We pickle/unpickle a *Iot*, so changing everything to pickle will be a noticeable performance regression.

The futurize library doesn't currently magically import cPickle as pickle under python 2. (http://python-future.org/reference.html#limitations)

So I suggest doing the following wherever pickle is currently imported:
```python
try:
    import cPickle as pickle
except ImportError:
    import pickle
```

[the remark this answers is not in the document]
Comment (DailyDreaming) [reply src/toil/jobStores/abstractJobStore.py]: Cool!  That makes much more sense.  I've just made the correction.

**Comment by ejacox** [comment]
@ryanking, there is still one test failing transiently. I think it is out of date. There are probably a few other updates that need to be futurized, too. Can you pull the latest for this PR?

**Comment by ejacox** [comment]
@ryanking - GitHub did finally let me update the PR. Thank you for this PR. Are you still willing to direct this effort and do a PR for the next step in #1780?

**Comment by ryanking** [comment]
Thanks for finishing this @ejacox. I can keep working on it, but only a few hours a week right now. If there's a sense of urgency about getting this done, some help would be useful.

**Comment by ejacox** [comment]
I liked the way we collaborated on this part. Can you take the lead and lay out what needs to be done next? For collaborating, what do you think if we create a branch off of master and a PR where we can discuss progress.

**Comment by ryanking** [comment]
sgtm

**Pull request merged**